### PR TITLE
implicitly grow BytesMut; add BufMutExt::chain_mut

### LINKED ISSUE #170: BytesMut should implicitly grow the internal storage in its `BufMut` implementation.
This will bring the implementation in line with that of `Vec`. This is a breaking change.

Relates to #131, #77.

### PR DESCRIPTION
**Pull request by carllerche**
This brings `BytesMut` in line with `Vec<u8>` behavior.

**Note** this PR also fixes an existing bug in `BytesMut::bytes_mut` that exposes invalid slices.

In order to fix a test, `BufMutExt::chain_mut` is provided. Withou this,
it is not possible to chain two `&mut [u8]`.

Closes #170

### CODE REVIEW

File: src/buf/ext/mod.rs [hunk 1]
```diff
@@ -124,6 +124,32 @@ pub trait BufMutExt: BufMut {
     fn writer(self) -> Writer<Self> where Self: Sized {
         writer::new(self)
     }
+
+    /// Creates an adaptor which will chain this buffer with another.
```
Comment (hawkw) [inline src/buf/ext/mod.rs]: Super nitpicky, but IIRC this spelling of "adaptor" is the British English spelling. Elsewhere in the docs, we use American English spellings, so I think it should be:
```suggestion
    /// Creates an adapter which will chain this buffer with another.
```

File: src/bytes_mut.rs [hunk 2]
```diff
@@ -924,20 +924,24 @@ impl Buf for BytesMut {
...
-        assert!(new_len <= self.cap);
+        assert!(new_len <= self.cap, "new_len = {}; capacity = {}", new_len, self.cap);
         self.len = new_len;
     }
 
     #[inline]
     fn bytes_mut(&mut self) -> &mut [mem::MaybeUninit<u8>] {
+        if self.capacity() == self.len() {
+            self.reserve(64);
+        }
```
Comment (hawkw) [inline src/bytes_mut.rs]: Does there need to be a check here that `self.capacity() != usize::MAX`, to avoid an overflow?

File: src/bytes_mut.rs [hunk 3]
```diff
@@ -924,20 +924,24 @@ impl Buf for BytesMut {
...
+
         unsafe {
-            slice::from_raw_parts_mut(self.ptr.as_ptr().offset(self.len as isize) as *mut mem::MaybeUninit<u8>, self.cap)
+            slice::from_raw_parts_mut(self.ptr.as_ptr().offset(self.len as isize) as *mut mem::MaybeUninit<u8>, self.cap - self.len)
```
Comment (hawkw) [inline src/bytes_mut.rs]: Nit/TIOLI: this expression is quite long, maybe there should be a `let`? Something like
```rust
            let ptr = self.ptr.as_ptr().offset(self.len as isize) as *mut mem::MaybeUninit<u8>;
            slice::from_raw_parts_mut(ptr, self.cap - self.len)
```

**Comment by carllerche** [comment]
@hawkw the bug was only on master for a bit (never released).

File: src/bytes_mut.rs (hunk 2 above)
Comment (carllerche) [reply src/bytes_mut.rs]: If there is a need, it would be in the `reserve` function.

File: src/bytes_mut.rs (hunk 3 above)
Comment (carllerche) [reply src/bytes_mut.rs]: Done.

File: src/bytes_mut.rs (hunk 2 above)
Comment (carllerche) [reply src/bytes_mut.rs]: I checked and there were issues. I have resolved them and added a test.

**Pull request merged**